### PR TITLE
NIAD-1517 Refactoring required to enable multiple GPC providers

### PR DESCRIPTION
**Pull request by mattd-kainos**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -190,6 +190,26 @@ The properties RequestId and Ssp-TraceID may only be populated if the log
 is emitted by application code. These are blank for logs emitted by framework or 
 third party code.
 
+## Mapping External (adaptor) URLs to internal (NHSD) URLs
+
+The adaptor proxies its requests to internal NHSD URLs via the Spine Secure Proxy.
+
+```
+Example: Get structured record for a patient at practive with ODS code A12345
```
Comment (andrewwa-kainos) [inline README.md:198]:
```suggestion
Example: Get structured record for a patient at practice with ODS code A12345
```

File: README.md [hunk 2]
```diff
@@ -190,6 +190,26 @@ The properties RequestId and Ssp-TraceID may only be populated if the log
...
+
+The request URL made to the adaptor includes the ODS code of the patient's practice. The two 
+variables (with * as wildcard) must have values matching the full request URL.
+
+https://gpcconsumer.prod.mydomain.com/A12345/STU3/1/gpconnect/structured/fhir/Patient/$gpc.getstructuredrecord
+|            $GPC_CONSUMER_URL      |          $GPC_CONSUMER_GPC_STRUCTURED_PATH                               |
+
+The adaptor then performs and SDS lookup and constructs a new path using the Spine
```
Comment (andrewwa-kainos) [inline README.md:206]:
```suggestion
The adaptor then performs an SDS lookup and constructs a new path using the Spine
```

File: service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/UrlsInResponseBodyRewriteFunction.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+@Slf4j
+@Component
+public class UrlsInResponseBodyRewriteFunction implements RewriteFunction<String, String> {
+    @Value("${gpc-consumer.gpc.gpcConsumerUrl}")
+    private String gpcConsumerUrl;
+    @Value("${gpc-consumer.gpc.overrideGpcProviderUrl}")
+    private String overrideGpcProviderUrl;
+
+    public static String replaceUrl(String gpcConsumerUrl, String overrideGpcProviderUrl, String responseBody) {
+        LOGGER.info(String.format("Replace host: %s, to: %s", overrideGpcProviderUrl, gpcConsumerUrl));
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/UrlsInResponseBodyRewriteFunction.java:21]:
```suggestion
        LOGGER.info("Replace host: {}, to: {}", overrideGpcProviderUrl, gpcConsumerUrl);
```

File: service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/UrlsInResponseBodyRewriteFunction.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        LOGGER.info("Replace host: {}, to: {}", overrideGpcProviderUrl, gpcConsumerUrl);
+        return responseBody.replace(overrideGpcProviderUrl, gpcConsumerUrl);
+    }
+
+    @Override
+    public Publisher<String> apply(ServerWebExchange serverWebExchange, String s) {
+        // TODO: Steps for NIAD-1283 to enable multiple GPC providers
+        // 1) Determine the correct prefix of *this* service - GPC_URL env var
+        // 2) Calculate a pattern for the URL we need to replace (based on previous SDS lookup, put onto event context)
+        // 3) Perform the replace operation using calculated pattern
```
Comment (Broom94) [inline service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/UrlsInResponseBodyRewriteFunction.java:30]: Are these comments still needed?

**Review by Broom94 — CHANGES_REQUESTED** [review]
Something strange seems to be happening with requests. For example request I send to gpcc:
http://localhost:8090/GP0001/STU3/1/gpconnect/structured/fhir/Patient/$gpc.getstructuredrecord

request it sends to wiremock/orangetestlabs etc has missing 'structured' or 'documents'

File: service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/UrlsInResponseBodyRewriteFunction.java (hunk 4 above)
Comment (mattd-kainos) [reply service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/UrlsInResponseBodyRewriteFunction.java:30]: They are what need to be done in the next subtask / to complete the story.

**Comment by mattd-kainos** [comment]
> Something strange seems to be happening with requests. For example request I send to gpcc:
> http://localhost:8090/GP0001/STU3/1/gpconnect/structured/fhir/Patient/$gpc.getstructuredrecord
> 
> request it sends to wiremock/orangetestlabs etc has missing 'structured' or 'documents'
> 
> steps to reproduce:
> 
> * run adaptor locally against wiremock
> * postman request against gpcc
> * sds enabled='true', sds url=wiremock

This is the problem with the SDS API Sandbox - it is missing this part of the URL (See https://gpc-structured-1-5-0.netlify.app/integration_spine_directory_service.html for what it should be returning)

Going to mitigate this with mocks: https://gpitbjss.atlassian.net/browse/NIAD-1533

File: service/src/main/resources/application.yml [hunk 5]
```diff
@@ -25,12 +25,12 @@ gpc-consumer:
     rootCA: ${GPC_CONSUMER_SPINE_ROOT_CA_CERT:}
     subCA: ${GPC_CONSUMER_SPINE_SUB_CA_CERT:}
     gpcConsumerUrl: ${GPC_CONSUMER_URL:http://localhost:8090}
-    gpcUrl: ${GPC_CONSUMER_GPC_GET_URL:}
+    overrideGpcProviderUrl: ${GPC_CONSUMER_OVERRIDE_GPC_PROVIDER_URL:}
```
Comment (Broom94) [inline service/src/main/resources/application.yml:28]: will we need separate gp2gp docker-compose pr for this variable update **GPC_CONSUMER_OVERRIDE_GPC_PROVIDER_URL**  and one below **GPC_CONSUMER_SSP_URL**?

**Review by Broom94 — APPROVED** [review]
Actually it doesn't matter until release

**Review by msolecka — APPROVED** [review]

**Pull request merged**